tikv-client/src/main/java/com/pingcap/tikv/predicates/TiKVScanAnalyzer.java

```diff
@@ -305,62 +317,176 @@ TiKVScanPlan buildIndexScan(
-          endKey = endKey.next();
+          signedEndKey = signedEndKey.next();
         }
       }
     } else {
       throw new TiClientInternalException("Empty access conditions");
     }
-    return new Pair<>(startKey, endKey);
+    return new ScanRange(signedStartKey, signedEndKey, null, null);
+  }
+
+  private ScanRange buildTableScanUnsignedKeyRangePerId(long id, IndexRange ir) {
+    Key signedStartClosedKey;
+    Key signedEndOpenKey;
+    Key unsignedStartClosedKey;
+    Key unsignedEndOpenKey;
+
+    if (ir.hasAccessKey()) {
+      checkArgument(
+          !ir.hasRange(), "Table scan must have one and only one access condition / point");
+      Key key = ir.getAccessKey();
+      checkArgument(key instanceof TypedKey, "Table scan key range must be typed key");
+      BigDecimal value = (BigDecimal) ((TypedKey) key).getValue();
+      if (value.compareTo(MAX_SIGNED_LONG) > 0) {
+        unsignedStartClosedKey = RowKey.toRowKey(id, new IntHandle(value.longValue()));
+        unsignedEndOpenKey = unsignedStartClosedKey.nextPrefix();
+        return new ScanRange(null, null, unsignedStartClosedKey, unsignedEndOpenKey);
+      } else {
+        signedStartClosedKey = RowKey.toRowKey(id, new IntHandle(value.longValue()));
+        signedEndOpenKey = signedStartClosedKey.nextPrefix();
+        return new ScanRange(signedStartClosedKey, signedEndOpenKey, null, null);
+      }
+    } else if (ir.hasRange()) {
+      Range<TypedKey> r = ir.getRange();
+      // Convert the range to low inclusive and high exclusive range
+      BigDecimal startClosedKey;
+      if (r.hasLowerBound()) {
+        BigDecimal startKey = (BigDecimal) r.lowerEndpoint().getValue();
+        startClosedKey =
+            r.lowerBoundType().equals(BoundType.CLOSED)
+                ? startKey
+                : (startKey.equals(MAX_UNSIGNED_LONG)
+                    ? BEYOND_UNSIGNED_LONG
+                    : startKey.add(BigDecimal.ONE));
+      } else {
+        startClosedKey = new BigDecimal(0);
+      }
+      boolean startClosedKeyBeyondUint64 = startClosedKey.equals(BEYOND_UNSIGNED_LONG);
+      if (startClosedKeyBeyondUint64) {
+        throw new IllegalArgumentException(
+            "Invalid range, startKey has exceeded MAX_UNSIGNED_LONG");
+      }
+
+      BigDecimal endOpenKey;
+      if (r.hasUpperBound()) {
+        BigDecimal endKey = (BigDecimal) r.upperEndpoint().getValue();
+        endOpenKey =
+            r.upperBoundType().equals(BoundType.OPEN)
+                ? endKey
+                : (endKey.equals(MAX_UNSIGNED_LONG)
+                    ? BEYOND_UNSIGNED_LONG
+                    : endKey.add(BigDecimal.ONE));
+      } else {
+        endOpenKey = BEYOND_UNSIGNED_LONG;
+      }
+      boolean endOpenKeyBeyondUint64 = Objects.equals(endOpenKey, BEYOND_UNSIGNED_LONG);
+
+      // [startKey, endOpenKey) is in range of [0, MAX_INT64], the same as [0, MAX_INT64 + 1)
+      if (!endOpenKeyBeyondUint64
```

Comment:
what if `[startKey, endOpenKey) is in range of [0, MAX_INT64]` and endOpenKeyBeyondUint64 is true

Reply:
There is already a else branch for throwing exception

Reply:
ok